packages/shylu/shylu_dd/frosch/src/SchwarzPreconditioners/FROSch_TwoLevelBlockPreconditioner_def.hpp

```diff
@@ -0,0 +1,269 @@
+                GDSWCoarseOperatorPtr gDSWCoarseOperator = Teuchos::rcp_static_cast<GDSWCoarseOperator<SC,LO,GO,NO> >(CoarseOperator_);
+                if (0>gDSWCoarseOperator->initialize(dimension,dofsPerNodeVec,repeatedNodesMapVec,dofsMapsVec,dirichletBoundaryDofsVec,nodeListVec)) ret -=10;
+            }
+            else if (!this->ParameterList_->get("CoarseOperator Type","IPOUHarmonicCoarseOperator").compare("RGDSWCoarseOperator")) {
+                this->ParameterList_->sublist("RGDSWCoarseOperator").sublist("CoarseSolver").sublist("MueLu").set("Dimension",(int)dimension);
+                RGDSWCoarseOperatorPtr rGDSWCoarseOperator = Teuchos::rcp_static_cast<RGDSWCoarseOperator<SC,LO,GO,NO> >(CoarseOperator_);
+                if (0>rGDSWCoarseOperator->initialize(dimension,dofsPerNodeVec,repeatedNodesMapVec,dofsMapsVec,dirichletBoundaryDofsVec,nodeListVec)) ret -=10;
+            }
+            else {
+                FROSCH_ASSERT(0!=0,"CoarseOperator Type unkown.");
+            }
+        }
+        return ret;
+    }
+    
+    template <class SC,class LO,class GO,class NO>
+    int TwoLevelBlockPreconditioner<SC,LO,GO,NO>::compute()
+    {
+        int ret = 0;
+
+        if (0>this->OverlappingOperator_->compute()) ret -= 1;
+        if (this->ParameterList_->get("TwoLevel",true)) {
+            if (0>CoarseOperator_->compute()) ret -= 10;
+        }
+        return ret;
+    }
+    
+    template <class SC,class LO,class GO,class NO>
+    void TwoLevelBlockPreconditioner<SC,LO,GO,NO>::describe(Teuchos::FancyOStream &out,
+                                                   const Teuchos::EVerbosityLevel verbLevel) const
+    {
+        FROSCH_ASSERT(0!=0,"describe() has be implemented properly...");
+    }
+    
+    template <class SC,class LO,class GO,class NO>
+    std::string TwoLevelBlockPreconditioner<SC,LO,GO,NO>::description() const
+    {
+        return "GDSW Preconditioner";
+    }
+    
+    template <class SC,class LO,class GO,class NO>
+    int TwoLevelBlockPreconditioner<SC,LO,GO,NO>::resetMatrix(CrsMatrixPtr &k)
+    {
+        this->K_ = k;
+        this->OverlappingOperator_->resetMatrix(this->K_);
+        if (this->ParameterList_->get("TwoLevel",true)) {
+            CoarseOperator_->resetMatrix(this->K_);
+            if (this->UseMultiplicative_) this->MultiplicativeOperator_->resetMatrix(this->K_);
+        }
+        return 0;
+    }
+    
+    template <class SC,class LO,class GO,class NO>
+    int TwoLevelBlockPreconditioner<SC,LO,GO,NO>::preApplyCoarse(MultiVectorPtr &x,MultiVectorPtr &y)
+    {
+        if (this->UseMultiplicative_) {
+            this->MultiplicativeOperator_->preApplyCoarse(*x,*y);
+        }
+        else{
+            FROSCH_ASSERT(false,"preApplyCoarse(MultiVectorPtr &x) only implemented for MultiplicativeOperator.")
+        }
+        return 0;
+    }
+    
+}
+
+#endif
```

Comment:
Most of this is the same as in the TwoLevelPreconditioner, right?

Reply:
yes, only for block systems.